File: src/main/java/game/Field.java
package game;

import java.awt.Dimension;
import java.awt.Graphics;
import java.awt.Point;
import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Random;

public class Field {
	
	private static final boolean SUPER_SAFE_FIRST_CLICK = false;
	
	private final int width;
	private final int height;
	private final int totalMines;
	private int minesLeft;
	private boolean created;
	private boolean lost;
	private boolean cheatMode;
	private Block blocks[][];
	
	public Field(int width, int height, int mines, boolean cheatMode) {
		this.width = width;
		this.height = height;
		this.totalMines = mines;
		blocks = new Block[width][height];
		this.cheatMode = cheatMode;
		reset();
	}
	
	public void reset() {
		lost = false;
		minesLeft = totalMines;
		for (int i = 0; i < width; i++) {
			for (int j = 0; j < height; j++) {
				blocks[i][j] = new Block(false);
			}
		}
		created = false;
	}

	public synchronized void paint(Graphics g, Dimension size) {
		final double pixelW = (double)size.width / width;
		final double pixelH = (double)size.height / height;
		for (int i = 0; i < width; i++) {
			for (int j = 0; j < height; j++) {
				blocks[i][j].paint(g, lost, cheatMode,
						(int)(i * pixelW),
						(int)(j * pixelH),
						(int)((i + 1) * pixelW) - (int)(i * pixelW),
						(int)((j + 1) * pixelH) - (int)(j * pixelH));
			}
		}
	}
	
	public void FlagBlock(Point point, Dimension dimension) {
		synchronized (blocks) {
			FlagBlockDirect(pointToCoord(point, dimension));
		}
	}
	
	public void FlagBlockDirect(Point point) {
		synchronized (blocks) {
			if (!blocks[point.x][point.y].isClear) {
				if (blocks[point.x][point.y].isFlagged) {
					blocks[point.x][point.y].isFlagged = false;
					minesLeft++;
				}
				else {
					blocks[point.x][point.y].isFlagged = true;
					minesLeft--;
				}
			}
		}
	}

	public void changeBlock(Point point, Dimension dimension) {
		synchronized (blocks) {
			changeBlockDirect(pointToCoord(point, dimension));
		}
	}
	
	public void changeBlockDirect(Point point) {
		synchronized (blocks) {
			if (!created) {
				created = true;
				plantMines(point);
			}
			clearBlock(point.x, point.y);
		}
	}
	
	public boolean checkWon() {
		for (int i = 0; i < width; i++) {
			for (int j = 0; j < height; j++) {
				if (!blocks[i][j].isClear && !blocks[i][j].isMine) {
					return false;
				}
			}
		}
		return true;
	}
	
	public boolean checkLost() {
		for (int i = 0; i < width; i++) {
			for (int j = 0; j < height; j++) {
				if (blocks[i][j].isClear && blocks[i][j].isMine) {
					return true;
				}
			}
		}
		return false;
	}
	
	public BlockData[][] getBoard() {
		synchronized (blocks) {
			BlockData[][] bd = new BlockData[width][height];
			for (int i = 0; i < width; i++) {
				for (int j = 0; j < height; j++) {
					bd[i][j] = blocks[i][j].getBlockData();
				}
			}
			return bd;
		}
	}

	public int getMines() {
		return totalMines;
	}

	public int getMinesLeft() {
		synchronized (blocks) {
			return minesLeft;
		}
	}
	
	private Point pointToCoord(Point p, Dimension size) {
		
		final double pixelW = (double)size.width / width;
		final double pixelH = (double)size.height / height;
		
		int x = (int) (p.x / pixelW);
		int y = (int) (p.y / pixelH);
		
		return new Point(x, y);
	}
	
	private void plantMines(Point point) {
		
		if (SUPER_SAFE_FIRST_CLICK) {
			superSafeSetup(point);
		}
		else {
			safeSetup(point);
		}
		
		for (int i = 0; i < width; i++) {
			for (int j = 0; j < height; j++) {
				if (blocks[i][j].isMine) {
					blocks[i][j].setNearbyMines(0);
				}
				else {
					int neighbors = 0;
					int cellXMax = Math.min(i + 2, width);
					int cellYMax = Math.min(j + 2, height);
					for (int cellX = Math.max(i - 1, 0); cellX < cellXMax; cellX++) {
						for (int cellY = Math.max(j - 1, 0); cellY < cellYMax; cellY++) {
							if (blocks[cellX][cellY].isMine) {
								neighbors++;
							}
						}	
					}
					blocks[i][j].setNearbyMines(neighbors - (blocks[i][j].isMine ? 1 : 0));
				}
			}
		}
	}

	private void safeSetup(Point point) {
		ArrayList<Point> possiblePoints = new ArrayList<Point>();
		for (int i = 0; i < width; i++) {
			for (int j = 0; j < height; j++) {
				if (i != point.x || j != point.y) {
					possiblePoints.add(new Point(i, j));
				}
			}
		}
		Random rand = new Random();
		for (int i = 0; i < totalMines; i++) {
			int index = rand.nextInt(possiblePoints.size());
			blocks[possiblePoints.get(index).x][possiblePoints.get(index).y].setMine(true);
			possiblePoints.remove(index);
		}
	}

	private void superSafeSetup(Point point) {
		ArrayList<Point> possiblePoints = new ArrayList<Point>();
		for (int i = 0; i < width; i++) {
			for (int j = 0; j < height; j++) {		
				if (Math.abs(i - point.x) > 1 || Math.abs(j - point.y) > 1) {
					possiblePoints.add(new Point(i, j));
				}
			}
		}
		Random rand = new Random();
		for (int i = 0; i < totalMines; i++) {
			int index = rand.nextInt(possiblePoints.size());
			blocks[possiblePoints.get(index).x][possiblePoints.get(index).y].setMine(true);
			possiblePoints.remove(index);
		}
	}
	
	private void clearBlock(int x, int y) {
		if (blocks[x][y].isClear) {
			int flagged = 0;
			int cellXMax = Math.min(x + 2, width);
			int cellYMax = Math.min(y + 2, height);
			for (int cellX = Math.max(x - 1, 0); cellX < cellXMax; cellX++) {
				for (int cellY = Math.max(y - 1, 0); cellY < cellYMax; cellY++) {
					if (blocks[cellX][cellY].isFlagged) {
						flagged++;
					}
				}	
			}
			if (flagged == blocks[x][y].getNearbyMines()) {
				for (int cellX = Math.max(x - 1, 0); cellX < cellXMax; cellX++) {
					for (int cellY = Math.max(y - 1, 0); cellY < cellYMax; cellY++) {
						if (!blocks[cellX][cellY].isClear) {
							clearBlock(cellX, cellY);
						}
					}	
				}
			}
		}
		else if (!blocks[x][y].isFlagged) {
			if (blocks[x][y].clear()) {
				lost = true;
			}
			else if (blocks[x][y].getNearbyMines() == 0) {
				int cellXMax = Math.min(x + 2, width);
				int cellYMax = Math.min(y + 2, height);
				for (int cellX = Math.max(x - 1, 0); cellX < cellXMax; cellX++) {
					for (int cellY = Math.max(y - 1, 0); cellY < cellYMax; cellY++) {
						if (!blocks[cellX][cellY].isClear) {
							clearBlock(cellX, cellY);
						}
					}
				}
			}
		}
	}

	public void save(long totalTime) {
		try (BufferedWriter bw = new BufferedWriter(new FileWriter(totalTime + ".txt"))) {
			bw.write("Time in milliseconds: " + totalTime + "\n");
			for (int i = 0; i < width; i++) {
				for (int j = 0; j < height; j++) {
					if (blocks[i][j].isMine) {
						bw.write('M');
					}
					else {
						bw.write('-');
					}
				}
				bw.write('\n');
			}
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
}

File: src/main/kotlin/game/Field.kt
package game

import forAll
import game.BlockState.*
import java.awt.Dimension
import java.awt.Graphics
import java.awt.Point
import java.io.File
import kotlin.math.abs
import kotlin.random.Random

class Field(
	private val config: GameConfig,
	private val cheatMode: Boolean
) {
	companion object {
		private const val SUPER_SAFE_FIRST_CLICK = false
	}

	private var minesLeft = 0
	private lateinit var state: FieldState
	private val blocks: Array<Array<Block>> = Array(config.width) { Array(config.height) { Block() } }

	val board: Array<Array<BlockData>>
		get() = synchronized(blocks) {
			Array(config.width) { i ->
				Array(config.height) { j ->
					blocks[i][j].blockData
				}
			}
		}

	init {
		reset()
	}

	fun reset() {
		minesLeft = config.mines
		blocks.forAll { it.reset() }
		state = FieldState.RESET
	}

	@Synchronized
	fun paint(g: Graphics, size: Dimension) {
		val pixelW = size.width.toDouble() / config.width
		val pixelH = size.height.toDouble() / config.height
		forEachPoint { i, j ->
			blocks[i][j].paint(
				g,
				state == FieldState.LOST,
				cheatMode,
				(i * pixelW).toInt(),
				(j * pixelH).toInt(),
				pixelW.toInt(),
				pixelH.toInt()
			)
		}
	}

	fun flagBlock(point: Point) = synchronized(blocks) {
		when (blocks[point.x][point.y].state) {
			HIDDEN -> {
				blocks[point.x][point.y].state = FLAGGED
				minesLeft--
			}
			FLAGGED -> {
				blocks[point.x][point.y].state = HIDDEN
				minesLeft++
			}
			REVEALED -> {}
		}
	}

	fun changeBlock(point: Point) = synchronized(blocks) {
		if (state == FieldState.RESET) {
			state = FieldState.GENERATED
			plantMines(point)
		}
		clearBlock(point.x, point.y)
	}

	fun checkWon() = blocks.all { row ->
		row.all { block ->
			block.isMine || block.state === REVEALED
		}
	}

	fun checkLost() = blocks.any { row ->
		row.any { block ->
			block.isMine && block.state == REVEALED
		}
	}

	fun getMinesLeft() = synchronized(blocks) {
		minesLeft
	}

	fun save(totalTime: Long) {
		val sb = StringBuffer("Time in milliseconds: $totalTime\n")

		for (i in 0 until config.width) {
			for (j in 0 until config.height) {
				sb.append(when(blocks[i][j].isMine) {
					true -> 'M'
					false -> '-'
				})
			}
			sb.append('\n')
		}

		val file = File("$totalTime.txt")
		file.writeText(sb.toString())
	}

	private fun plantMines(initialPoint: Point) {
		val possibleMines = when {
			SUPER_SAFE_FIRST_CLICK -> getSuperSafeMines(initialPoint)
			else -> getSafeMines(initialPoint)
		}.toMutableList()

		for (i in 0 until config.mines) {
			val point = possibleMines.removeAt(
				Random.nextInt(possibleMines.size)
			)
			blocks[point.x][point.y].run {
				isMine = true
				state = HIDDEN
				nearbyMines = 0.toByte()
			}
		}

		forEachPoint { i, j ->
			if (blocks[i][j].isMine) {
				forEachAround(i, j) { x, y ->
					blocks[x][y].nearbyMines++
				}
			}
		}
	}

	private fun getSafeMines(initialPoint: Point): List<Point> {
		return filterPoints { i, j ->
			i != initialPoint.x || j != initialPoint.y
		}
	}

	private fun getSuperSafeMines(initialPoint: Point): List<Point> {
		return filterPoints { i, j ->
			abs(i - initialPoint.x) > 1 || abs(j - initialPoint.y) > 1
		}
	}

	private fun clearBlock(i: Int, j: Int) {
		when (blocks[i][j].state) {
			FLAGGED -> {}
			REVEALED -> {
				var flagged = 0
				forEachAround(i, j) { x, y ->
					if (blocks[x][y].state == FLAGGED) {
						flagged++
					}
				}
				if (flagged == blocks[i][j].nearbyMines.toInt()) {
					forEachAround(i, j) { x, y ->
						if (blocks[x][y].state == HIDDEN) {
							clearBlock(x, y)
						}
					}
				}
			}
			HIDDEN -> {
				blocks[i][j].state = REVEALED
				if (blocks[i][j].isMine) {
					state = FieldState.LOST
				} else if (blocks[i][j].nearbyMines.toInt() == 0) {
					forEachAround(i, j) { x, y ->
						if (blocks[x][y].state == HIDDEN) {
							clearBlock(x, y)
						}
					}
				}
			}
		}
	}

	private fun filterPoints(predicate: (i: Int, j: Int) -> Boolean): List<Point> {
		return (0 until config.width).flatMap { i ->
			(0 until config.height).filter { j ->
				predicate(i, j)
			}.map { j ->
				Point(i, j)
			}
		}
	}

	private fun forEachPoint(action: (i: Int, j: Int) -> Unit) {
		(0 until config.width).forEach { i ->
			(0 until config.height).forEach { j ->
				action(i, j)
			}
		}
	}

	private fun forEachAround(i: Int, j: Int, action: (x: Int, y: Int) -> Unit) {
		for (x in (i - 1).coerceAtLeast(0) until (i + 2).coerceAtMost(config.width)) {
			for (y in (j - 1).coerceAtLeast(0) until (j + 2).coerceAtMost(config.height)) {
				action(x, y)
			}
		}
	}
}